lib/execution_segment_test.go

```diff
@@ -96,4 +184,133 @@ func TestExecutionSegmentSplit(t *testing.T) {
 	assert.Equal(t, "7/16:1/2", segments[3].String())
 }
 
-//TODO: a lot of other tests
+func TestExecutionSegmentScale(t *testing.T) {
+	var es = new(ExecutionSegment)
+	require.NoError(t, es.UnmarshalText([]byte("0.5")))
+	require.Equal(t, int64(1), es.Scale(2))
+	require.Equal(t, int64(2), es.Scale(3))
+
+	require.NoError(t, es.UnmarshalText([]byte("0.5:1.0")))
+	require.Equal(t, int64(1), es.Scale(2))
+	require.Equal(t, int64(1), es.Scale(3))
+}
+
+func TestExecutionSegmentCopyScaleRat(t *testing.T) {
+	var es = new(ExecutionSegment)
+	var twoRat = big.NewRat(2, 1)
+	var threeRat = big.NewRat(3, 1)
+	require.NoError(t, es.UnmarshalText([]byte("0.5")))
+	require.Equal(t, oneRat, es.CopyScaleRat(twoRat))
+	require.Equal(t, big.NewRat(3, 2), es.CopyScaleRat(threeRat))
+
+	require.NoError(t, es.UnmarshalText([]byte("0.5:1.0")))
+	require.Equal(t, oneRat, es.CopyScaleRat(twoRat))
+	require.Equal(t, big.NewRat(3, 2), es.CopyScaleRat(threeRat))
+
+	var nilEs *ExecutionSegment
+	require.Equal(t, twoRat, nilEs.CopyScaleRat(twoRat))
+	require.Equal(t, threeRat, nilEs.CopyScaleRat(threeRat))
+}
+
+func TestExecutionSegmentInPlaceScaleRat(t *testing.T) {
+	var es = new(ExecutionSegment)
+	var twoRat = big.NewRat(2, 1)
+	var threeRat = big.NewRat(3, 1)
+	var threeSecondsRat = big.NewRat(3, 2)
+	require.NoError(t, es.UnmarshalText([]byte("0.5")))
+	require.Equal(t, oneRat, es.InPlaceScaleRat(twoRat))
+	require.Equal(t, oneRat, twoRat)
+	require.Equal(t, threeSecondsRat, es.InPlaceScaleRat(threeRat))
+	require.Equal(t, threeSecondsRat, threeRat)
+
+	es = stringToES(t, "0.5:1.0")
+	twoRat = big.NewRat(2, 1)
+	threeRat = big.NewRat(3, 1)
+	require.Equal(t, oneRat, es.InPlaceScaleRat(twoRat))
+	require.Equal(t, oneRat, twoRat)
+	require.Equal(t, threeSecondsRat, es.InPlaceScaleRat(threeRat))
+	require.Equal(t, threeSecondsRat, threeRat)
+
+	var nilEs *ExecutionSegment
+	twoRat = big.NewRat(2, 1)
+	threeRat = big.NewRat(3, 1)
+	require.Equal(t, big.NewRat(2, 1), nilEs.InPlaceScaleRat(twoRat))
+	require.Equal(t, big.NewRat(2, 1), twoRat)
+	require.Equal(t, big.NewRat(3, 1), nilEs.InPlaceScaleRat(threeRat))
+	require.Equal(t, big.NewRat(3, 1), threeRat)
+}
+
```

Comment:
I think all of the tests here should be run with `t.Parallel()`.

Also, one of the tests I planned to add here would have been the ultimate validation of https://github.com/loadimpact/k6/issues/997. Basically, in a loop of say, 100 times, generate a random number of random fractions with a sum of `1`. Then, turn these into execution segments (checking for errors), and use the resulting segments to `Scale()` another bunch of random integer numbers. Then we check if the sum of the scaled nubmers is equal to the starting number.

Reply:
Though maybe we should merge this PR and simply add that test directly in #1007?

Reply:
I also wanted to have all the tests (including all the executors) run through list of executionSegments. , maybe not random (although that sounds like a good idea). But decided to go for ...  well ... writing some amount of tests for most of the code and hopefully getting some functions in place ready for doing bigger and more complicated tests. Also in the case of the executors the math started to get too hairy for me ;) 

I would prefer if we merge this HUGE PR in the other HUGEST PR and than possibly add smaller and more targeted tests ... specifically so that reviewing those changes doesn't take a day ... or three ;)